### PR TITLE
Patch ZeChrale's raids

### PR DESCRIPTION
**Pull request by MonkeyTedBaker**
Fixed bugs present in ZeChrales/Monocle raids branch. Sends out notifications for raids. Webhook type is 'raid' instead of 'pokemon'. Also supports web hook notifications.

Tested and running fine with > 30,000rpm of hashkeys at full capacity.

_Fixes applied only on backend._

~I could test about 3 hours today before the raids ended in my region.~
~Out of 18k+ raids found, 2k were left with pokemon_id unupdated (eggs unhatched) in DB.~
~Not sure yet if it was a bug or just my insufficient workers failing to pass by the raid a second time.~

Checked the results today. There's no problem with updating hatched raid eggs. It only fails to update when there are not enough workers passing by the gym.

====

Added raid icons in `monocle-icons`. Since it is not yet merged, it is needed to do the following to load icons temporarily from my fork.

====

To load icons (for testing front end) from MonkeyTedBaker repo.
```
git config --file=.gitmodules submodule.monocle-icons.url https://github.com/MonkeyTedBaker/monocle-icons.git
git config --file=.gitmodules submodule.monocle-icons.branch patch-raids
git submodule sync
```

To revert monocle-icons to Noctem repo,
```
git config --file=.gitmodules submodule.monocle-icons.url https://github.com/Noctem/monocle-icons.git
git config --file=.gitmodules submodule.monocle-icons.branch master
git submodule sync
```

### CODE REVIEW

File: monocle/sanitized.py [hunk 1]
```diff
@@ -197,6 +198,7 @@
     'PLAYER_LOCALE': {'country': 'US', 'language': 'en', 'timezone': 'America/Denver'},
     'PROVIDER': None,
     'PROXIES': None,
+	'RAIDS_FILTER': (1, 2, 3, 4, 5),
```
Comment (evenly-epic-mule) [inline monocle/sanitized.py]: tab

and I'd guess filtering on (3,4,5) by default would be better

File: monocle/static/css/main.css [hunk 2]
```diff
@@ -146,13 +146,19 @@
 .pokemarker {
     position: absolute;
     left: -16px;
-    top: -20px;
+    top: -16px;
     width: 32px;
+	height: 32px;
```
Comment (evenly-epic-mule) [inline monocle/static/css/main.css]: tab

File: monocle/static/js/main.js [hunk 3]
```diff
@@ -69,8 +88,9 @@ function monitor (group, initial) {
 
 monitor(overlays.Pokemon, false)
 monitor(overlays.Trash, true)
+monitor(overlays.Raids, true)
 monitor(overlays.Gyms, true)
-monitor(overlays.Workers, false)
+monitor(overlays.Workers, true)
```
Comment (evenly-epic-mule) [inline monocle/static/js/main.js]: why change this?

File: monocle/worker.py [hunk 4]
```diff
@@ -1264,9 +1269,33 @@ def normalize_gym(raw):
             'prestige': raw.gym_points,
```
Comment (evenly-epic-mule) [inline monocle/worker.py]: is that still provided or can it be removed?
note: if you remove it there may be places where it is accessed without checking the availability

**Review by evenly-epic-mule — CHANGES_REQUESTED** [review]
I haven't tested it but some small things I noticed

File: monocle/db.py [hunk 5]
```diff
@@ -578,7 +669,8 @@ def _get_forts(session):
             fs.guard_pokemon_id,
             fs.last_modified,
             f.lat,
-            f.lon
+            f.lon,
+            f.slots_available
```
Comment (Zoruk) [inline monocle/db.py]: `f.slots_available` -> `fs.slots_available`

File: monocle/sanitized.py (hunk 1 above)
Comment (phonezawphyo) [inline monocle/sanitized.py]: Will set 3,4,5 as suggested

File: monocle/static/js/main.js (hunk 3 above)
Comment (phonezawphyo) [inline monocle/static/js/main.js]: accident xD

File: monocle/worker.py (hunk 4 above)
Comment (phonezawphyo) [inline monocle/worker.py]: Will leave it for legacy reasons. Somebody should make a separate PR to remove prestige.

File: monocle/db.py (hunk 5 above)
Comment (phonezawphyo) [inline monocle/db.py]: Thanks. Lifesaver.

File: monocle/sanitized.py (hunk 1 above)
Comment (MonkeyTedBaker) [reply monocle/sanitized.py]: Will set 3,4,5 as suggested

File: monocle/db.py (hunk 5 above)
Comment (MonkeyTedBaker) [reply monocle/db.py]: Thanks. Lifesaver.

File: monocle/worker.py (hunk 4 above)
Comment (MonkeyTedBaker) [reply monocle/worker.py]: Will leave it for legacy reasons. Somebody should make a separate PR to remove prestige.

File: monocle/static/js/main.js (hunk 3 above)
Comment (MonkeyTedBaker) [reply monocle/static/js/main.js]: accident xD

File: monocle/web_utils.py [hunk 6]
```diff
@@ -118,6 +118,31 @@ def get_pokemarkers(after_id=0):
...
 
+def get_raid_markers(names=POKEMON):
+    with session_scope() as session:
+        markers = []
+        raids = session.query(Raid) \
+            .filter(Raid.time_end > time())
+        for raid in raids:
+            fort_id = raid.fort_id
+            fort = session.query(Fort) \
+                .filter(Fort.external_id == fort_id) \
```
Comment (Zoruk) [inline monocle/web_utils.py]: `Fort.external_id == fort_id` => `Fort.id == fort_id`

**Comment by Zoruk** [comment]
Can you also provide the icons? They're not in the official icons repo

File: monocle/web_utils.py (hunk 6 above)
Comment (MonkeyTedBaker) [reply monocle/web_utils.py]: Thanks.

**Comment by MonkeyTedBaker** [comment]
@Zoruk I only intend to implement backend (as I don't use the frontend myself). However, I've added raid icons for convenience of testers. Pls refer to the description at the top for instructions to load raid icons for testing. Thank you.

File: monocle/static/css/main.css [hunk 7]
```diff
@@ -146,13 +146,19 @@
...
     width: 32px;
+    height: 32px;
     text-align: center;
     margin: 0;
     padding: 0;
 }
 
+.raid_egg {
+    width: 30px;
+    height: 30x;
```
Comment (tmagnier) [inline monocle/static/css/main.css:159]: height: 30**p**x;

**Comment by MrCanan** [comment]
Is it possible to add the "WebHook" feature to allow receiving egg and raid notifications on PokeAlarm ? =)

**Comment by MrCanan** [comment]
Can you add Raids to webhook to be able to have notifications ? =)

**Comment by MonkeyTedBaker** [comment]
There you go. Raid notifications.

**Comment by mfaalk** [comment]
Would it be possible to show all defending pokémons instead of just the last one entered at the frontend?
It would be great if you can display CP of the pokémon and maybe even trainer name.

**Comment by MrCanan** [comment]
Hi MonkeyTedBaker,

Thanks to have add Raids Notifications. =)

But their is an error with PokeAlarm webhook : 
`2017-08-02 16:04:53,239 [    "niceraids"][   Manager][   ERROR] Encountered error during processing: TypeError: unsupported operand type(s) for -: 'NoneType' and 'datetime.datetime'`

It seems their is a missing parameter to send to webhook : `raid_begin`, so normally simply add `data['message']['raid_begin'] = pokemon['time_spawn']` just before the line `'data['message']['disappear_time'] = ts`
And the line : `'data['message']['disappear_time'] = ts` should be replace with : `'data['message']['raid_end'] = ts`
But, you will understand this problem and find a solution faster than me. =)

Very cordially.

**Comment by MonkeyTedBaker** [comment]
@MrCanan Since it is adding only 2 extra parameters, I've added it to the raid web hook as you've suggested.
If the web hook format is much more different than this, I would advise you to ask for PokeAlarm to support this format.
Haven't tested it with PokeAlarm since I don't use it.

@mfaalk Yes but not in this branch. Im doing it in another branch of mine, gymplus. It's still heavily in WIP and not recommended to pull it.

**Comment by MonkeyTedBaker** [comment]
For those who would enjoy extra gym info for this version of raids, I've made another pull request with all the juicy details #350

**Comment by MonkeyTedBaker** [comment]
Better version offered in #350

**Pull request closed without merging**